internal/runtimehandlerhooks/high_performance_hooks_linux.go

```diff
@@ -631,69 +640,103 @@ func (h *HighPerformanceHooks) setIRQLoadBalancing(ctx context.Context, c *oci.C
 		return fmt.Errorf("find container %s CPUs", c.ID())
 	}
 
-	newIRQBalanceSetting, err := h.updateNewIRQSMPAffinityMask(lspec.Resources.CPU.Cpus, enable)
+	h.updateIRQSMPAffinityLock.Lock()
+	defer h.updateIRQSMPAffinityLock.Unlock()
+
+	newIRQBalanceSetting, err := h.updateNewIRQSMPAffinityMask(ctx, c.ID(), c.Name(), lspec.Resources.CPU.Cpus, enable)
 	if err != nil {
 		return err
 	}
-
-	isIrqConfigExists := fileExists(h.irqBalanceConfigFile)
-
-	if isIrqConfigExists {
-		if err := h.updateIrqBalanceConfigFile(newIRQBalanceSetting); err != nil {
-			return err
-		}
+	// Now, restart the irqbalance service or run irqbalance --oneshot command.
+	// On failure, this will log errors but will not return them, as it is not critical for the pod to start.
+	if !h.handleIRQBalanceRestart(ctx, c.Name()) {
+		h.handleIRQBalanceOneShot(ctx, c.Name(), newIRQBalanceSetting)
 	}
 
-	if !isServiceEnabled(irqBalancedName) || !isIrqConfigExists {
-		if _, err := exec.LookPath(irqBalancedName); err != nil {
-			// irqbalance is not installed, skip the rest; pod should still start, so return nil instead
-			log.Warnf(ctx, "Irqbalance binary not found: %v", err)
-
-			return nil
-		}
-		// run irqbalance in daemon mode, so this won't cause delay
-		cmd := cmdrunner.Command(irqBalancedName, "--oneshot")
-		additionalEnv := irqBalanceBannedCpus + "=" + newIRQBalanceSetting
-		cmd.Env = append(os.Environ(), additionalEnv)
+	return nil
+}
 
-		return cmd.Run()
+// handleIRQBalanceRestart handles the restart of the irqbalance service.
+func (h *HighPerformanceHooks) handleIRQBalanceRestart(ctx context.Context, cName string) bool {
+	// If the irqbalance service is enabled, restart it and return.
+	// systemd's StartLimitBurst might cause issues here when container restarts occur in very
+	// quick succession and the parameter must be reconfigured for this to work correctly.
+	// See:
+	// https://github.com/cri-o/cri-o/pull/8834/commits/b96928dcbb7956e0ebde42238e88955831411216
+	if !isServiceEnabled(irqBalancedName) || !fileExists(h.irqBalanceConfigFile) {
+		return false
 	}
 
+	log.Debugf(ctx, "Container %q restarting irqbalance service", cName)
+
 	if err := restartIrqBalanceService(); err != nil {
 		log.Warnf(ctx, "Irqbalance service restart failed: %v", err)
+
+		return false
```

Comment:
I introduced a small logic change here: if systemctl restart irqbalance fails, the logic now falls back to irqbalance --oneshot